runtime/compiler/optimizer/InterpreterEmulator.hpp

```diff
@@ -0,0 +1,272 @@
+ *        can't be reasoned about are represented by a dummy operand \ref _unknownOperand
+ *        which doesn't carry any extra information. Currently, with state mode only
+ *        supports methodhandle thunk archetypes.
+ */
+
+#ifndef INTERPRETER_EMULATOR_INCL
+#define INTERPRETER_EMULATOR_INCL
+
+#include "il/Block.hpp"
+#include "ilgen/ByteCodeIteratorWithState.hpp"
+#include "ilgen/J9ByteCodeIterator.hpp"
+#include "compile/Compilation.hpp"
+#include "optimizer/Inliner.hpp"
+#include "optimizer/J9Inliner.hpp"
+#include "optimizer/J9EstimateCodeSize.hpp"
+#include "env/TRMemory.hpp"
+
+class IconstOperand;
+class KnownObjOperand;
+class MutableCallsiteTargetOperand;
+
+class Operand
+   {
+   public:
+      TR_ALLOC(TR_Memory::EstimateCodeSize);
+      virtual IconstOperand *asIconst(){ return NULL;}
+      virtual KnownObjOperand *asKnownObject(){ return NULL;}
+      virtual MutableCallsiteTargetOperand* asMutableCallsiteTargetOperand(){ return NULL;}
+      virtual bool isUnkownOperand(){ return true;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return TR::KnownObjectTable::UNKNOWN;}
+   };
+
+/*
+ * \class KnownOperand
+ *
+ * \brief represent operands that can be reasoned about at compile time
+ */
+class KnownOperand : public Operand
+   {
+   public:
+      TR_ALLOC(TR_Memory::EstimateCodeSize);
+      virtual bool isUnkownOperand(){ return false;}
+   };
+
+class IconstOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::EstimateCodeSize);
+      IconstOperand (int x): intValue(x) { }
+      virtual IconstOperand *asIconst() { return this;}
+      int32_t intValue;
+   };
+
+class KnownObjOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::EstimateCodeSize);
+      KnownObjOperand(TR::KnownObjectTable::Index koi):knownObjIndex(koi){ }
+      virtual KnownObjOperand *asKnownObject(){ return this;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return knownObjIndex;}
+      TR::KnownObjectTable::Index knownObjIndex;
+   };
+
+/*
+ * \class MutableCallsiteTargetOperand
+ *
+ * \note This class is used to support mutable callsite because both the methodhandle object
+ *       and the mutable callsite needs to be tracked so that when creating \c TR_J9MutableCallSite
+ *       the mutable callsite object can be set for the callsite even though it's really the
+ *       methodhandle object that's on the operand stack.
+ *
+ * \see getReturnValueForInvokevirtual
+ * \see refineResolvedCalleeForInvokestatic
+ * \see visitInvokestatic
+ */
+class MutableCallsiteTargetOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::EstimateCodeSize);
+      MutableCallsiteTargetOperand (TR::KnownObjectTable::Index methodHandleIndex, TR::KnownObjectTable::Index mutableCallsiteIndex):
+         methodHandleIndex(methodHandleIndex),
+         mutableCallsiteIndex(mutableCallsiteIndex){}
```

Comment:
it seems odd that this does not delegate to super in any way... should it? You add two fields even though the parent already has one which seems a bit wasteful...

Reply:
changed the object to inherit `KnownObjOperand` and made `methodHandleIndex` reuse `knownObjIndex` field of `KnownObjOperand`. Also added method `getMethodHandleIndex` and `getMutableCallsiteIndex ` to keep the readability.